python_modules/libraries/dagster-dbt/dagster_dbt/cli/utils.py

```diff
@@ -111,74 +136,85 @@ def execute_cli(
-                        ),
-                    )
-                    log_level = json_line.get(
-                        # Attempt to get the log level from the dbt-core==1.3.* format
-                        "level",
-                        # Otherwise, try to get the message from the dbt-core==1.4.* format
-                        json_line.get("info", {}).get(
-                            "level",
-                            # If all else fails, default to the `debug` level
-                            "debug",
-                        ),
-                    )
-        elif "Done." not in line:
-            # attempt to parse a log level out of the line
-            if "ERROR" in line:
-                log_level = "error"
-            elif "WARN" in line:
-                log_level = "warn"
-
-        messages.append(message)
-        if capture_logs:
-            log.log(coerce_valid_log_level(log_level), message)
+        cli_msg = _parse_line(line, json_log_format)
 
-    process.wait()
-    return_code = process.returncode
+        if cli_msg.message is not None:
+            messages.append(cli_msg.message)
 
-    log.info("dbt exited with return code {return_code}".format(return_code=return_code))
+        # yield the parsed values
+        yield cli_msg
 
-    raw_output = "\n\n".join(raw_output)
+    process.wait()
+    return_code = process.returncode
 
     if return_code == 2:
         raise DagsterDbtCliFatalRuntimeError(messages=messages)
 
     if return_code == 1 and not ignore_handled_error:
         raise DagsterDbtCliHandledRuntimeError(messages=messages)
 
+    yield DbtCliEvent(return_code=return_code)
+
+
+def execute_cli(
+    executable: str,
+    command: str,
+    flags_dict: Mapping[str, Any],
+    log: Any,
+    warn_error: bool,
+    ignore_handled_error: bool,
+    target_path: str,
+    docs_url: Optional[str] = None,
+    json_log_format: bool = True,
+    capture_logs: bool = True,
+) -> DbtCliOutput:
+    """Executes a command on the dbt CLI in a subprocess."""
+    check.str_param(executable, "executable")
+    check.str_param(command, "command")
+    check.mapping_param(flags_dict, "flags_dict", key_type=str)
+    check.bool_param(warn_error, "warn_error")
+    check.bool_param(ignore_handled_error, "ignore_handled_error")
+
+    command_list = _create_command_list(
+        executable=executable,
+        warn_error=warn_error,
+        json_log_format=json_log_format,
+        command=command,
+        flags_dict=flags_dict,
+    )
+    log.info(f"Executing command: {' '.join(command_list)}")
+
+    return_code = 0
+    lines, parsed_json_lines = [], []
+    for event in execute_cli_stream(
+        command_list=command_list,
+        json_log_format=json_log_format,
+        ignore_handled_error=ignore_handled_error,
+    ):
+        if event.return_code is not None:
+            return_code = event.return_code
+            log.info(f"dbt exited with return code {return_code}")
+            break
```

Comment:
This feels like its own distinct event from `DbtCliOutput`, since it is a terminal event. I suggest that the `execute_cli_stream` be executed on the process, and then at the very end, the return code can be directly taken from the process, outside of the iteration.

Reply:
Agree that return_code is pretty distinct from the other DbtCliEvent stuff, so I'm down to remove it. However, the next PR in the stack adds another callsite for execute_cli_stream, which increases the need to have everything factored together. That is, we need to do a few things:

1. create a subprocess with some specific arguments / env vars
2. iterate through the stdout of that subprocess, yielding parsed events as we go
3. wait for the subprocess to complete, raising an error (in some cases) based off the return code
4. have the calling function aware of that return code value (so that it can construct a DbtCliOutput)

So we're stuck in this situation where we want the calling function to consume all the yielded events, but also know the return code. The only way we can do this without having some function which handles both yielding DbtCliEvents and return codes is if the calling process is given the responsibility for 1, 3, and 4. These could take the form of two functions (_get_dbt_subprocess() and _cleanup_dbt_subprocess()), but I figured that might be more confusing than anything. So I just updated the return type of `execute_cli_stream` to be Union[DbtCliEvent, int].